File: web/player/src/validate.js
// Shared schema is bundled at build time. Validation checks recorded data only.
const TRACE_SCHEMA = __TRACE_SCHEMA__;
const DEFAULT_MAX_BYTES = 128 * 1024 * 1024;
const own = (o,k) => Object.prototype.hasOwnProperty.call(o,k);
function requireData(ok, path, message) { if (!ok) throw new Error(`${path}: ${message}`); }
function same(a,b) {
  if (a === b) return true;
  if (!a || !b || typeof a !== 'object' || typeof b !== 'object') return false;
  const keys = Object.keys(a);
  return keys.length === Object.keys(b).length && keys.every(k => own(b,k) && same(a[k],b[k]));
}
// Implements exactly the JSON Schema keywords emitted by tools/build_schemas.py.
function checkSchema(v, s, path='$') {
  if (s.$ref) return checkSchema(v,TRACE_SCHEMA.$defs[s.$ref.split('/').pop()],path);
  if (s.anyOf) {
    requireData(s.anyOf.some(option => {try {checkSchema(v,option,path);return true;} catch {return false;}}),path,'invalid value'); return;
  }
  if (own(s,'const')) requireData(v===s.const,path,`expected ${s.const}`);
  if (s.enum) requireData(s.enum.includes(v),path,'unsupported value');
  if (s.type) {
    const valid = s.type==='null' ? v===null : s.type==='array' ? Array.isArray(v) : s.type==='object' ? v!==null && typeof v==='object' && !Array.isArray(v) : s.type==='integer' ? Number.isSafeInteger(v) : typeof v===s.type;
    requireData(valid,path,`expected ${s.type}`);
  }
  if (typeof v==='number') {
    requireData(Number.isFinite(v),path,'expected finite number');
    if (own(s,'minimum')) requireData(v>=s.minimum,path,`minimum ${s.minimum}`);
    if (own(s,'maximum')) requireData(v<=s.maximum,path,`maximum ${s.maximum}`);
    if (own(s,'exclusiveMinimum')) requireData(v>s.exclusiveMinimum,path,`must exceed ${s.exclusiveMinimum}`);
  }
  if (typeof v==='string') {
    if (s.minLength) requireData(v.length>=s.minLength,path,'string too short');
    if (s.maxLength) requireData([...v].length<=s.maxLength,path,'string too long');
  }
  if (Array.isArray(v)) {
    if (s.minItems) requireData(v.length>=s.minItems,path,'array too short');
    if (s.maxItems) requireData(v.length<=s.maxItems,path,'array too long');
    if (s.items) v.forEach((x,i)=>checkSchema(x,s.items,`${path}[${i}]`));
  } else if (v && typeof v==='object') {
    for (const k of s.required || []) requireData(own(v,k),path,`missing ${k}`);
    for (const k of Object.keys(v)) {
      if (s.properties && own(s.properties,k)) checkSchema(v[k],s.properties[k],`${path}.${k}`);
      else if (s.additionalProperties===false) throw new Error(`${path}: unknown field ${k}`);
      else if (s.additionalProperties && typeof s.additionalProperties==='object') checkSchema(v[k],s.additionalProperties,`${path}.${k}`);
    }
  }
}
function indexBy(list,key,path) {
  const map=new Map();
  for (const x of list) {requireData(!map.has(x[key]),path,`duplicate ${x[key]}`);map.set(x[key],x);}
  return map;
}
function near(a,b,path) { requireData(a!==null && Math.abs(a-b)<=Math.max(1e-6,1e-9*Math.max(Math.abs(a),Math.abs(b))),path,'recorded metric mismatch'); }
function validateTrace(data) {
  requireData(data?.format==='satnet-edu.trace' && data.schema_version==='1.0.0','$','supported format satnet-edu.trace version 1.0.0');
  checkSchema(data,TRACE_SCHEMA);
  const s=data.experiment.scenario, r=data.recording, objects=indexBy(data.objects,'id','objects'), queries=indexBy(r.route_queries,'id','queries');
  const scenarioObjects=indexBy(s.objects,'id','scenario.objects'), shells=indexBy(s.constellations,'id','constellations'), slots=new Set();
  requireData(/(?:Z|\+00:00)$/.test(s.epoch_utc) && Number.isFinite(Date.parse(s.epoch_utc)) && +s.epoch_utc.slice(0,4)>=1957 && +s.epoch_utc.slice(0,4)<=2056,'epoch_utc','expected UTC, year 1957–2056');
  for (const k of ['epoch_utc','name','seed']) requireData(s[k]===data.experiment[k],`experiment.${k}`,'scenario mismatch');
  requireData(same(s.link_policy,data.model.link_policy),'model.link_policy','scenario mismatch');
  requireData(objects.size===scenarioObjects.size && objects.size<=s.limits.max_nodes,'objects','count mismatch/limit');
  for (const [id,o] of objects) {
    requireData(same(o,scenarioObjects.get(id)),`objects.${id}`,'scenario mismatch');
    const fields=o.kind==='satellite'?['tle','orbit']:['lat_deg','lon_deg'];
    requireData(fields.every(k=>own(o,k)),`objects.${id}`,'missing object definition');
    const forbidden=o.kind==='satellite'?['lat_deg','lon_deg']:['tle','orbit','constellation_id','plane','slot'];
    requireData(forbidden.every(k=>!own(o,k)),`objects.${id}`,'incompatible fields');
    if (own(o,'constellation_id')) {
      const c=shells.get(o.constellation_id), key=JSON.stringify([o.constellation_id,o.plane,o.slot]);
      requireData(c && own(o,'plane') && own(o,'slot') && o.plane<c.planes && o.slot<c.sats_per_plane && !slots.has(key),`objects.${id}`,'invalid/duplicate slot'); slots.add(key);
    } else requireData(!own(o,'plane') && !own(o,'slot'),`objects.${id}`,'slot needs constellation');
  }
  for (const [id,c] of shells) requireData(s.objects.filter(o=>o.constellation_id===id).length===c.planes*c.sats_per_plane,`constellations.${id}`,'missing slots');
  requireData(r.capabilities.routes===Boolean(queries.size),'capabilities.routes','query mismatch');
  for (const q of queries.values()) requireData(objects.has(q.source)&&objects.has(q.target),'queries','unknown endpoint');
  requireData(r.end_s>=r.start_s,'recording','reversed interval');
  const count=Math.floor((r.end_s-r.start_s)/r.step_s);
  requireData(count+1<=s.limits.max_samples,'recording','sample limit exceeded');
  const times=new Set([r.start_s,r.end_s]);
  for (let i=1;i<=count;i++) if(r.start_s+i*r.step_s<r.end_s) times.add(r.start_s+i*r.step_s);
  const events=[];
  for (const f of s.failures) {
    requireData(objects.has(f.node_id)&&f.end_s>f.start_s,'failures','invalid interval/node');
    for (const [k,kind] of [['start_s','failure_start'],['end_s','failure_end']]) if(f[k]>=r.start_s&&f[k]<=r.end_s) {times.add(f[k]);events.push([f[k],f.node_id,kind]);}
  }
  requireData(times.size<=s.limits.max_samples && data.frames.length===r.sample_count && same(data.frames.map(f=>f.t_s),[...times].sort((a,b)=>a-b)),'frames.t_s','sample grid mismatch');
  requireData(same(events.map(JSON.stringify).sort(),data.events.map(e=>JSON.stringify([e.t_s,e.node_id,e.kind])).sort()),'events','scheduled event mismatch');
  data.frames.forEach((f,i)=>{
    const p=`frames[${i}]`, states=indexBy(f.node_states,'id',p), links=indexBy(f.links,'id',p), routes=indexBy(f.routes,'query_id',p), pairs=new Set();
    requireData(states.size===objects.size&&[...states.keys()].every(k=>objects.has(k)),p,'missing/unknown node state');
    requireData(routes.size===queries.size&&[...routes.keys()].every(k=>queries.has(k)),p,'missing/unknown query result');
    for (const [id,state] of states) {
      const scheduled=s.failures.some(fault=>fault.node_id===id&&fault.start_s<=f.t_s&&f.t_s<fault.end_s);
      requireData(state.enabled!==scheduled,p,'scheduled failure state mismatch');
    }
    for (const l of links.values()) {
      requireData(states.has(l.source)&&states.has(l.target)&&l.source!==l.target,p,'unknown endpoint/self loop');
      const pair=JSON.stringify([l.source,l.target].sort());requireData(!pairs.has(pair),p,'duplicate pair');pairs.add(pair);
      requireData(states.get(l.source).enabled&&states.get(l.target).enabled,p,'disabled link endpoint');
      const ground=[l.source,l.target].filter(k=>objects.get(k).kind==='ground_station').length;
      requireData(ground===(l.kind==='gsl'?1:0),p,'invalid link kind');near(l.propagation_ms,l.distance_km/data.model.c_km_s*1000,p);
    }
    for (const [qid,route] of routes) {
      const q=queries.get(qid), nodes=route.node_ids, lids=route.link_ids, disabled=!states.get(q.source).enabled||!states.get(q.target).enabled;
      if(route.status==='unreachable') {
        requireData(!nodes.length&&!lids.length&&['hops','distance_km','propagation_ms'].every(k=>route[k]===null),p,'unreachable needs empty path and null metrics');
        requireData(route.reason===(disabled?'endpoint_disabled':'no_path'),p,'invalid reason');continue;
      }
      requireData(!disabled&&route.reason===null&&nodes.length>0&&nodes[0]===q.source&&nodes.at(-1)===q.target&&new Set(nodes).size===nodes.length,p,'invalid path endpoints/cycle');
      requireData(lids.length===nodes.length-1&&route.hops===lids.length,p,'hop mismatch');
      requireData(nodes.every(n=>states.has(n)&&states.get(n).enabled),p,'unknown/disabled path node');
      requireData(nodes.slice(1,-1).every(n=>objects.get(n).kind==='satellite'),p,'GS cannot relay');
      lids.forEach((id,j)=>{const l=links.get(id);requireData(l&&((l.source===nodes[j]&&l.target===nodes[j+1])||(l.target===nodes[j]&&l.source===nodes[j+1])),p,'missing/mismatched path edge');});
      near(route.distance_km,lids.reduce((a,id)=>a+links.get(id).distance_km,0),p);
      near(route.propagation_ms,lids.reduce((a,id)=>a+links.get(id).propagation_ms,0),p);
    }
  });
  return data;
}

File: src/satnet_edu/trace/validate.py
"""Schema and semantic validation. Recorded paths are checked, never re-routed."""

import json
import math
from functools import lru_cache
from importlib.resources import files

from jsonschema import Draft202012Validator

from ..config import utc_epoch
from ..engine.clock import sample_times
from ..engine.orbit import phases


@lru_cache(maxsize=2)
def _validator(kind):
    return Draft202012Validator(
        json.loads(
            files("satnet_edu")
            .joinpath(f"assets/{kind}.schema.json")
            .read_text(encoding="utf-8")
        )
    )


def _finite(v, path="$"):
    if isinstance(v, float) and not math.isfinite(v):
        raise ValueError(f"{path}: expected finite number")
    if isinstance(v, dict):
        for k, x in v.items():
            _finite(x, f"{path}.{k}")
    if isinstance(v, (list, tuple)):
        for i, x in enumerate(v):
            _finite(x, f"{path}[{i}]")


def _schema(data, kind):
    _finite(data)
    error = next(_validator(kind).iter_errors(data), None)
    if error:
        raise ValueError(
            f"{kind}.{'.'.join(map(str, error.absolute_path))}: {error.message}"
        )


def _unique(items, key, path):
    result = {}
    for item in items:
        if item[key] in result:
            raise ValueError(f"{path}: duplicate {key} {item[key]}")
        result[item[key]] = item
    return result


def _require(ok, path, message):
    if not ok:
        raise ValueError(f"{path}: {message}")


def _close(a, b, path):
    _require(
        a is not None and math.isclose(a, b, rel_tol=1e-9, abs_tol=1e-6),
        path,
        f"expected {b}, got {a}",
    )


def validate_scenario(data, *, structural=True):
    if structural:
        _schema(data, "scenario")
    utc_epoch(data["epoch_utc"])
    objects = _unique(data["objects"], "id", "objects")
    constellations = _unique(data["constellations"], "id", "constellations")
    _require(
        len(objects) <= data["limits"]["max_nodes"], "objects", "exceeds max_nodes"
    )
    occupied = set()
    for sid, o in objects.items():
        path = f"objects.{sid}"
        fields = (
            ("tle", "orbit") if o["kind"] == "satellite" else ("lat_deg", "lon_deg")
        )
        _require(all(k in o for k in fields), path, f"requires {fields}")
        forbidden = (
            ("tle", "orbit", "constellation_id", "plane", "slot")
            if o["kind"] == "ground_station"
            else ("lat_deg", "lon_deg")
        )
        _require(not any(k in o for k in forbidden), path, "incompatible object fields")
        if "constellation_id" in o:
            c = constellations.get(o["constellation_id"])
            _require(
                c is not None and "plane" in o and "slot" in o,
                path,
                "invalid membership",
            )
            m = (c["id"], o["plane"], o["slot"])
            _require(
                o["plane"] < c["planes"]
                and o["slot"] < c["sats_per_plane"]
                and m not in occupied,
                path,
                "invalid or duplicate slot",
            )
            occupied.add(m)
            expected_raan, expected_phase = phases(
                o["plane"], o["slot"], c["planes"], c["sats_per_plane"]
            )
            _require(
                o["orbit"]
                == dict(
                    altitude_km=c["altitude_km"],
                    inclination_deg=c["inclination_deg"],
                    raan_deg=expected_raan,
                    phase_deg=expected_phase,
                ),
                path,
                "orbit differs from constellation definition",
            )
        else:
            _require(
                "plane" not in o and "slot" not in o,
                path,
                "slot requires constellation",
            )
    for cid, c in constellations.items():
        _require(
            sum(m[0] == cid for m in occupied) == c["planes"] * c["sats_per_plane"],
            f"constellations.{cid}",
            "missing slots",
        )
    for f in data["failures"]:
        _require(
            f["node_id"] in objects and f["end_s"] > f["start_s"],
            "failures",
            "unknown node or empty/reversed interval",
        )
    return data


def validate_trace(data):
    if (
        not isinstance(data, dict)
        or data.get("format") != "satnet-edu.trace"
        or data.get("schema_version") != "1.0.0"
    ):
        raise ValueError(
            "trace: supported format satnet-edu.trace, schema_version 1.0.0"
        )
    _schema(data, "trace")
    s = data["experiment"]["scenario"]
    validate_scenario(s, structural=False)
    for key in ("epoch_utc", "name", "seed"):
        _require(
            data["experiment"][key] == s[key], f"experiment.{key}", "scenario mismatch"
        )
    _require(
        data["model"]["link_policy"] == s["link_policy"],
        "model.link_policy",
        "scenario mismatch",
    )
    objects = _unique(data["objects"], "id", "objects")
    _require(
        objects == {o["id"]: o for o in s["objects"]}, "objects", "scenario mismatch"
    )
    recording = data["recording"]
    queries = _unique(recording["route_queries"], "id", "route_queries")
    _require(
        recording["capabilities"]["routes"] == bool(queries),
        "capabilities.routes",
        "query mismatch",
    )
    for q in queries.values():
        _require(
            q["source"] in objects and q["target"] in objects,
            "route_queries",
            "unknown endpoint",
        )
    times = sample_times(
        recording["start_s"],
        recording["end_s"] - recording["start_s"],
        recording["step_s"],
        s["failures"],
        s["limits"]["max_samples"],
    )
    _require(
        [f["t_s"] for f in data["frames"]] == times,
        "frames.t_s",
        "must match full sample grid including event boundaries",
    )
    _require(
        len(data["frames"]) == recording["sample_count"],
        "recording.sample_count",
        "frame count mismatch",
    )
    for i, f in enumerate(data["frames"]):
        p = f"frames[{i}]"
        states = _unique(f["node_states"], "id", p + ".node_states")
        links = _unique(f["links"], "id", p + ".links")
        routes = _unique(f["routes"], "query_id", p + ".routes")
        _require(states.keys() == objects.keys(), p, "missing/unknown node state")
        _require(routes.keys() == queries.keys(), p, "missing/unknown query result")
        for sid, state in states.items():
            scheduled = any(
                failure["node_id"] == sid
                and failure["start_s"] <= f["t_s"] < failure["end_s"]
                for failure in s["failures"]
            )
            _require(
                state["enabled"] != scheduled,
                p + ".node_states." + sid,
                "scheduled failure state mismatch",
            )
            xyz = state["position_ecef_km"]
            radius = math.sqrt(sum(v * v for v in xyz))
            _require(radius > 0, p + ".node_states." + sid, "zero ECEF position")
            _close(
                state["altitude_km"],
                radius - data["model"]["earth_radius_km"],
                p + ".altitude_km",
            )
            _close(
                state["lat_deg"],
                math.degrees(math.atan2(xyz[2], math.hypot(xyz[0], xyz[1]))),
                p + ".lat_deg",
            )
            longitude = math.degrees(math.atan2(xyz[1], xyz[0]))
            _require(
                abs((state["lon_deg"] - longitude + 180) % 360 - 180) < 1e-6,
                p + ".lon_deg",
                "ECEF mismatch",
            )
        pairs = set()
        for lid, l in links.items():
            a, b = l["source"], l["target"]
            _require(
                a in states and b in states and a != b,
                p + ".links",
                "unknown endpoint or self loop",
            )
            pair = tuple(sorted((a, b)))
            _require(pair not in pairs, p + ".links", "duplicate pair")
            pairs.add(pair)
            _require(
                states[a]["enabled"] and states[b]["enabled"],
                p + ".links",
                "disabled endpoint",
            )
            kinds = [objects[a]["kind"], objects[b]["kind"]]
            _require(
                kinds.count("ground_station") == (1 if l["kind"] == "gsl" else 0),
                p + ".links",
                "invalid link kind",
            )
            _close(
                l["propagation_ms"],
                l["distance_km"] / data["model"]["c_km_s"] * 1000,
                p + ".links." + lid,
            )
            _close(
                l["distance_km"],
                math.dist(states[a]["position_ecef_km"], states[b]["position_ecef_km"]),
                p + ".links." + lid + ".distance_km",
            )
        for qid, r in routes.items():
            rp = p + ".routes." + qid
            q = queries[qid]
            disabled = (
                not states[q["source"]]["enabled"] or not states[q["target"]]["enabled"]
            )
            nodes, lids = r["node_ids"], r["link_ids"]
            if r["status"] == "unreachable":
                _require(
                    not nodes
                    and not lids
                    and all(
                        r[k] is None for k in ("hops", "distance_km", "propagation_ms")
                    ),
                    rp,
                    "unreachable requires empty path and null metrics",
                )
                _require(
                    r["reason"] == ("endpoint_disabled" if disabled else "no_path"),
                    rp,
                    "invalid reason",
                )
                continue
            _require(
                not disabled and r["reason"] is None,
                rp,
                "reachable has disabled endpoint or reason",
            )
            _require(
                bool(nodes)
                and nodes[0] == q["source"]
                and nodes[-1] == q["target"]
                and len(set(nodes)) == len(nodes),
                rp,
                "invalid endpoints/cycle",
            )
            _require(
                len(lids) == len(nodes) - 1 and r["hops"] == len(lids),
                rp,
                "invalid hop count",
            )
            _require(
                all(n in states and states[n]["enabled"] for n in nodes),
                rp,
                "unknown/disabled path node",
            )
            _require(
                all(objects[n]["kind"] == "satellite" for n in nodes[1:-1]),
                rp,
                "GS cannot relay",
            )
            for j, lid in enumerate(lids):
                _require(
                    lid in links
                    and {links[lid]["source"], links[lid]["target"]}
                    == {nodes[j], nodes[j + 1]},
                    rp,
                    "missing/mismatched path edge",
                )
            _close(
                r["distance_km"],
                sum(links[l]["distance_km"] for l in lids),
                rp + ".distance_km",
            )
            _close(
                r["propagation_ms"],
                sum(links[l]["propagation_ms"] for l in lids),
                rp + ".propagation_ms",
            )
    expected = sorted(
        (f[k], f["node_id"], kind)
        for f in s["failures"]
        for k, kind in [("start_s", "failure_start"), ("end_s", "failure_end")]
        if recording["start_s"] <= f[k] <= recording["end_s"]
    )
    _require(
        sorted((e["t_s"], e["node_id"], e["kind"]) for e in data["events"]) == expected,
        "events",
        "scheduled events mismatch",
    )
    return data
